### PR TITLE
feat: Implement testing type switch promos

### LINKED ISSUE #26448: Add Testing Type switcher to Specs Explorer
Screenshot 2023-05-23 at 10 43 28 AM

Screenshot 2023-05-23 at 10 43 36 AM

Screenshot 2023-05-23 at 10 53 28 AM

At the top of the specs list, there will be two buttons, one for e2e and one for component testing. Each tab will have the following features:

- 	Icon for test testing type (always shown)
- 	Testing type label text
	- 	E2E Specs
	- 	Component Specs
- 	A question mark icon that is shown when a testing type is unconfigured.
- 	Active/inactive states

**When both testing types are configured**
Buttons act as a switcher to put Cypress into the other testing mode. 
There are no new side effects of this, the transition happens exactly as it does today when a user switches via the top-left button in the side nav - the browser closes, launchpad displays a spinner, and then the browser re-opens in the new testing type.

**When one testing type is not configured**
The button for the unconfigured type contains a question mark, and clicking it displays a promo in-place in the specs list page which gives a summary of the testing type and invites users to set it up.
This has a side effect: it records a "campaign viewed" event through the Cloud's /anon-collect  endpoint with the following values
- Component testing:
		campaign: Specs Switcher CT
		medium: Not set up
		_if a "CT eligible" framework and bundler are detected:_
		payload: { framework: {framework}, bundler: {bundler} }
		(see ComponentTestingBannerAvailable.vue  for example)
- End-to-End testing
		campaign: Specs Switcher E2E
		medium: Not set up
		payload: { framework: {framework}, bundler: {bundler} }

The implementation of the `Promo` elements for each unconfigured state are covered under tickets #26837 and #26836. This ticket should handle creation of the switch, switching when target type is configured, and replacing Specs List content with a placeholder when not configured.

### LINKED ISSUE #26836: Create "Setup E2E" Promo
When in CT mode without E2E configured, we should display a Promo to help users learn about and get started with E2E within the Specs Explorer.

The Testing Type switcher shown here will be implemented under #26448 , this ticket is for implementing the Promo.

**CTA Headline:** End-to-end Testing is not set up for this project
**CTA Text:** Try it out now, or read our guides to learn more 
  * "read our guides" is a link to [on.cypress.io/writing-first-test](http://on.cypress.io/writing-first-test)
  * clicking this creates a "campaign completed" event using the /machine-collect endpoint
		campaign: Read our guide
		medium: E2E Preview

**CTA Button Text:** Quick setup
  * clicking this creates a "campaign completed" event using the /machine-collect endpoint
		campaign: Quick setup
		medium: E2E Preview

**Body headline:** What is end-to-end testing?
**Body text:** End-to-end tests are perfect for testing your entire application from a user's perspective. Visit any URL and test critical user flows exactly as they work in production.
**Body list:**
- Item 1: Test full user journeys across multiple pages, using cy.visit() to load up a web page for your test.
- Item 2: Work with live API calls or mock them with cy.intercept()  to test any possible state of your app.
- Item 3: Use additional commands to manage sessions and multiple origins in your tests.

Screenshot 2023-05-23 at 11 11 59 AM

### LINKED ISSUE #26837: Create "Setup CT" Promo
When in E2E mode without CT configured, we should display a Promo to help users learn about and get started with CT within the Specs Explorer.

The Testing Type switcher shown here will be implemented under #26448 , this ticket is for implementing the Promo.

**CTA Headline:** Component Testing is not set up for this project
**CTA Text:** Try it out now, or read our guides to learn more
  * "read our guides" is a link to [on.cypress.io/component](http://on.cypress.io/component)
  * clicking this creates a "campaign completed" event using the /machine-collect endpoint
		campaign: Read our guide
		medium: CT Preview

**CTA Button Text:** Quick setup
  * clicking this creates a "campaign completed" event using the /machine-collect endpoint
		campaign: Quick setup
		medium: CT Preview

**Body headline:** What is Component Testing?
**Body text:** Cypress Component Testing is a technique for developers to write, test, and debug components in isolation without ever leaving the browser — and without having to run your entire application.
**Body list:**

- Item 1: Instead of visiting a page, use cy.mount() to render your front-end components directly in the browser.
- Item 2: Test from a user's perspective with all the same Cypress commands that are available in end-to-end testing.
- Item 3: To get started, you can generate the config for Angular, React, or Vue (and more!) in just a few clicks.
		* each framework name has an inline icon in front of it for that framework
		* the framework names should be on-links created in #26838
			* they will share the following params:
				utm_medium: CT Preview,
				utm_content: if detected `{framework}-{bundler}`, otherwise empty 
				utm_campaign: the content of the link that was clicked, so Angular, Vue, React or "more"
				for reference see: https://github.com/cypress-io/cypress/blob/develop/packages/app/src/specs/banners/ComponentTestingAvailableBanner.vue#L122
			* Each docs link should also record a "campaign completed" event
				medium: CT Preview
				campaign: the content of the link that was clicked, so Angular, Vue, React or "more"

**Feedback link:** Give feedback
  * Existing onlink: on.cypress.io/component-survey-q2-23
  * clicking this creates a "campaign completed" event using the /machine-collect endpoint
		campaign: Give feedback
		medium: CT Preview

Screenshot 2023-05-23 at 11 45 23 AM

### PR DESCRIPTION
**Pull request by mike-plummer**
- Closes #26448 
- Closes https://github.com/cypress-io/cypress/issues/26837
- Closes https://github.com/cypress-io/cypress/issues/26836

### Additional details
A new testing type switcher is being added to the Specs Explorer. It has the following behaviors:

### If other testing type is configured
* no "question mark" icon is displayed
* Clicking the other testing type immediately switches the active testing type (closes and reopens browser).

### If other testing type is not configured
* A question mark icon is displayed
* Clicking the other testing type hides the specs list and shows a Promo with informational content about that testing type.
* Promo will:
  - Record an anon event on display
  - Provide a "Quick setup" button that closes browser and sends user to setup flow for other testing type. Clicking button records a `machine-collect` event
  - Provides docs and framework links with appropriate UTM params

[Figma](https://www.figma.com/proto/J6uGNpRC2ct4wIq5j6JzBj/Specs-Page-v1.1-%40latest?page-id=2%3A126&type=design&node-id=579-49216&viewport=355%2C230%2C0.16&scaling=scale-down&starting-point-node-id=579%3A49214&show-proto-sidebar=1)

_Note:_ The following discrepancies with the Figma have been surfaced with design and approved:
1. Different style of Question Mark icon in switcher
2. Switcher is shorter (complies with DS styling)
3. No "skeleton" rows are displayed behind promo
4. There's some extra height from the SpecsList virtualist which causes the promo to center a bit low - going to see how Adam's changes to SpecsList styling impact this

### Steps to test
1. Create a project with CT and E2E set up
5. Open project using this branch, observe switcher displays with no question mark. Clicking each testing type should move Cypress into target testing type.
6. Close project. Edit config file and comment out "component" entry. Reopen project into E2E
7. Switcher should display question mark on CT mode. Click CT button and observe CT promo is displayed. Verify links include appropriate UTM params. Click Quick Setup and verify user is sent to launchpad setup flow for CT.
8. Close project. Edit config file - uncomment "component" and comment out "e2e" config. Reopen project into CT.
9. Switcher should display question mark on E2E mode. Click E2E button and observe E2E promo is displayed. Verify links include appropriate UTM params. Click Quick Setup and verify user is sent to E2E mode.

### How has the user experience changed?

**Both testing types configured**
Screenshot 2023-05-31 at 1 26 09 PM

**CT not configured**
Screenshot 2023-05-31 at 1 26 56 PM

**E2E not configured**
Screenshot 2023-05-31 at 1 27 53 PM

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by astone123** [comment]
A couple of questions before I dive into the code:
1. Should the testing type switcher persist state if the user navigates to a different page and then back to the specs list? Right now if you navigate to a testing type and see the promo, then navigate away and back to it, the switcher goes back to the current testing type. This makes intuitive sense to me, I just want to make sure that was intentional.
2. There's some weirdness in the UI:

After switching testing types once I saw it like this (this was immediately after I used the switcher to set up e2e testing, not sure if relevant):
Screenshot 2023-06-01 at 8 27 46 AM

Also, it looks like every time we switch testing types using the switcher, the selected UI testing type shows up during the loading state for the browser. For example (can't believe I screenshotted this fast enough)
Screenshot 2023-06-01 at 8 30 12 AM
This looks kind of strange to me. Anything we can do about that?

**Comment by mike-plummer** [comment]
@astone123 
1. Correct, there are no state persistence requirements for the switcher, so the expectation at present is that it will always default to the "active" testing type after a refresh or navigation
2. Weird. I'll take a look at the horizontal overlap issue, might be conflicting with the grid column widths we're using in the header
3. I saw that yesterday while demoing with Mark - I think there may be a hard z-index override on that component. I'll take a look

File: cli/CHANGELOG.md
Comment (astone123) [inline cli/CHANGELOG.md]: I think since this is a feature the next version should be `12.14.0`, right?

File: packages/app/cypress/e2e/specs_list_switcher.cy.ts
Comment (astone123) [inline packages/app/cypress/e2e/specs_list_switcher.cy.ts:1]: Can we add test cases here for when both testing types are configured?

File: packages/app/src/specs/TestingTypePromo.cy.tsx
Comment (astone123) [inline packages/app/src/specs/TestingTypePromo.cy.tsx:3]: Should we add assertions in here to make sure that the links have the correct UTM params? Looks like there's a lot of logic around that in the component. Might be useful to define expectations here

**Review by astone123 — COMMENTED** [review]
Just a few comments. Also I'm curious why the snapshot cache files were updated?

**Comment by mike-plummer** [comment]
> Just a few comments. Also I'm curious why the snapshot cache files were updated?

@astone123 kitchensink and unit tests are failing due to weird errors related to require & package.json. I ran the snapshot update task just to make sure it wasn't a snapshot mismatch since there are updates to the yarn.lock in this PR

**Comment by astone123** [comment]
Just tested again. I'm not seeing the issues that I saw before. One more thing, should the bottom of the promo have this much padding? Looks a little off to me

Screenshot 2023-06-01 at 11 56 59 AM

**Comment by mike-plummer** [comment]
> Just tested again. I'm not seeing the issues that I saw before. One more thing, should the bottom of the promo have this much padding? Looks a little off to me

@astone123 Removed extra padding in latest commit

**Comment by marktnoonan** [comment]
Still testing, here's an edge case I found - if I remove the component testing configuration while CT is open, we don't seem to handle that, and the UI gets to an undefined state (no "project/branch" item in the top left of the nav).

I guess ideally we'd put you back at the launchpad to choose your testing type again and either you open e2e or set up CT.

I haven't tested the current behavior, maybe this PR doesn't have anything to do with it.

Screenshot 2023-06-01 at 4 02 49 PM

File: packages/app/src/specs/TestingTypePromo.vue
Comment (marktnoonan) [inline packages/app/src/specs/TestingTypePromo.vue]:
```suggestion
      class="m-auto mt-[32px]"
```

If we set a top margin here we can get a little closer to the Figma spacing between the top bar and the and the promo itself. What do you think - is there some bad side effect of this that I'm missing?

Screenshot 2023-06-01 at 4 29 32 PM

**Comment by astone123** [comment]
@mike-plummer One more minor thing I'm seeing - when I launch a project that has e2e configured but not CT, when the app loads I see the CT promo flash on screen very quickly. Probably not a huge deal, but I imagine we could prevent that from happening for a smoother experience.

For example
https://www.loom.com/share/f40652ad01994d458c50bc52e880d7fc

**Comment by marktnoonan** [comment]
Overall this is looking really great. One question on the size of the full-width version, I couldn't put my finger on this yesterday when we talked about the button size.

In Figma the full-width variant is 858px. This produces less text wrapping and a good amount of space between the title and the button:

Screenshot 2023-06-01 at 4 42 58 PM

It looks like our promo component doesn't get bigger than 480px. Is this known/intentional at this point, based on other existing uses of `PromoCard`?

**Comment by mike-plummer** [comment]
> It looks like our promo component doesn't get bigger than 480px. Is this known/intentional at this point, based on other existing uses of `PromoCard`?

Current width for Promo was established for the Runs & Debug promos - we can customize it here if needed, but since this isn't a Design System component the styling is very use-case specific so far

**Comment by marktnoonan** [comment]
I think it makes sense to get this matching the width in the designs, we'd end up being more similar in size & appearance to the "with image" variant on the Debug page, which looks good to me.

Screenshot 2023-06-01 at 4 55 34 PM

**Comment by warrensplayer** [comment]
image

"Last Updated" has no padding from the tabs at narrow viewports.  Also, the width of the tabs causes the alignment of the columns to be off at certain widths.

**Comment by mike-plummer** [comment]
@marktnoonan Last commit should resolve the top margin and promo width issues you raised
@astone123 Last commit should prevent promo from flashing on reload
@warrensplayer Did not have a chance to look at the column widths thing - good eye, hadn't noticed that myself.

Someone will need to look into the column width thing above while I'm out. Also, the `Debug.cy.ts` spec has started failing at some point that I haven't had a chance to look into along with some weird build issues on kitchensink.

**Comment by warrensplayer** [comment]
image

I am not sure what the final design decisions.  Looking at the Figma, it depicts a skeleton state behind the promo.  If this iteration is not going to do that, then I think we can just hide the skeleton column headers. As it stands now, they look out of place to me.  Thoughts?

**Comment by mike-plummer** [comment]
@warrensplayer The skeleton header was one of the iterative designs - the "skeleton rows behind the promo" make it look a lot like a modal which causes some UX issues so we're holding off on that for now (see PR description). I'll defer to @marktnoonan and design re: whether the skeleton header makes sense

File: packages/app/src/specs/TestingTypePromo.vue
Comment (warrensplayer) [reply packages/app/src/specs/TestingTypePromo.vue]: The issue with the vertical spacing is how the containing div for the promo is interacting with its siblings and parent. There is an `h-full` (i.e. `height: 100%`) on the containing div which makes the div for the promo be the same height as the entire lower right grid cell.  In the branch for the notifications that Adam has been working on, the SpecsList page has been refactored to use an inner grid to control the different sections of that part of the page.  It makes things much easier to work with.

**Comment by warrensplayer** [comment]
> It looks like our promo component doesn't get bigger than 480px. Is this known/intentional at this point, based on other existing uses of PromoCard?

@marktnoonan I think the Promo could have a wide single column version.  I do think that the design for this promo in Figma is too wide however.  Text line lengths scan better when they are not so wide. I left a comment in Figma asking about this.

File: packages/app/src/specs/TestingTypePromo.vue
Comment (warrensplayer) [reply packages/app/src/specs/TestingTypePromo.vue]: I have a suggestion for how to fix this for now and will add a comment to the code.

File: packages/app/src/components/promo/PromoCard.vue
Comment (warrensplayer) [inline packages/app/src/components/promo/PromoCard.vue]: 870px seems too wide for readability for a line of text.  
```suggestion
    classes.push('xl:grid-cols-[640px]')
```

File: packages/app/src/specs/SpecsList.vue
Comment (warrensplayer) [inline packages/app/src/specs/SpecsList.vue]:
```suggestion
    <TestingTypePromo
      v-if="testingType.showTestingTypePromo.value"
      class="spec-list-container p-[32px] overflow-y-auto"
      :testing-type="testingType.viewedTestingType.value"
      @activate-testing-type="testingType.activateTestingType"
    />
```

File: packages/app/src/specs/TestingTypePromo.vue
Comment (warrensplayer) [inline packages/app/src/specs/TestingTypePromo.vue]:
```suggestion
  <div class="flex">
    <Promo
      :campaign="campaign"
      medium="Not set up"
      :instance-id="promoInstanceId"
      class="m-auto"
    >
```

**Review by warrensplayer — COMMENTED** [review]
Suggestions for centering the content for the promo in the box and getting the container to scroll like the specs list does rather than scroll the whole section.

File: packages/app/src/components/promo/PromoCard.vue
Comment (marktnoonan) [reply packages/app/src/components/promo/PromoCard.vue]: I updated this based on Chris's feedback, Figma had become incorrect 👍

File: packages/app/src/specs/TestingTypePromo.vue
Comment (marktnoonan) [reply packages/app/src/specs/TestingTypePromo.vue]: I removed the `h-full`, but I think keeping the top margin is more accurate, the design doesn't seem to call for it to be vertically centered in the space, just down a little from the top.

**Comment by marktnoonan** [comment]
Thanks for the suggestions @warrensplayer and @astone123. Here's the latest responsive behavior and positioning:

https://github.com/cypress-io/cypress/assets/8340719/6e2883fe-6a16-4f5a-a668-fa54153a9217

I removed the skeleton header since it's no longer in the latest Figma. I did spend some time on the main body skeleton, but there's actually a bit of complexity involved in keeping it looking good across screen sizes. So aside from UX concerns, it's going to take a bit more time to polish it and keep everything aligned.

This catches the PR up to all feedback I'm aware of.

**Comment by astone123** [comment]
@marktnoonan looks good overall, but I'm still seeing this behavior that I mentioned in [this comment](https://github.com/cypress-io/cypress/pull/26894#issuecomment-1572750047). Not a huge issue but it would be nice if it was an easy change

Edit: Actually what I'm seeing now is that the **e2e promo** is flashing on screen (which is the testing type that is configured)

File: packages/app/src/composables/useTestingType.ts
Comment (astone123) [inline packages/app/src/composables/useTestingType.ts:37]: We probably can revert this line and just check for `!isE2EConfigured.value`, etc. since we'll know by this point that it is defined. Maybe it's good to be explicit though

**Review by astone123 — APPROVED** [review]
I just tested again, looks good!

**Comment by astone123** [comment]
Fixed this issue in https://github.com/cypress-io/cypress/pull/26894/commits/707350f1f44bf23062529309c54db963263b10c3

https://www.loom.com/share/abac6ed229394f90b5cf5b923ce27034

**Comment by lmiller1990** [comment]
I *think* I still saw the flashing banner bug, but I realized it was another banner. I am unsure if it was introduced here, but I suspect it was not. I'll find out.

Here's the bug recorded. It only happens sometimes, the first time you open a project?


https://github.com/cypress-io/cypress/assets/19196536/03e12430-6331-49e6-9890-aec103960a0f

**Comment by lmiller1990** [comment]
The actual feature works great, I tested it out, looks good!

**Comment by lmiller1990** [comment]
I re-ran `yarn` and I can no longer reproduce the flicking bug I found on either this branch or develop. I'm not sure why it happened, but I don't think it was introduced here at any rate.

Edit, maybe I'd missed https://github.com/cypress-io/cypress/pull/26894/commits/54cb1a0fa45b7d4bc7a03c30f026539416a40f6f where @marktnoonan fixed the issue. It seems fine now!

**Comment by lmiller1990** [comment]
I pushed a single commit after getting it all green that removed a superfluous snapshot (this component is captured in several others): https://github.com/cypress-io/cypress/pull/26894/commits/59df96a4cf7f6a691f1556a0687cc259d4de90a6.

**Pull request merged**
